backend/dataall/modules/shares_base/db/share_object_state_machines.py

```diff
@@ -0,0 +1,296 @@
+import logging
+
+from dataall.base.db import exceptions
+from dataall.modules.shares_base.services.shares_enums import (
+    ShareObjectActions,
+    ShareObjectStatus,
+    ShareItemActions,
+    ShareItemStatus,
+)
+from dataall.modules.shares_base.db.share_object_repositories import ShareObjectRepository
+
+logger = logging.getLogger(__name__)
+
+
+class Transition:
+    def __init__(self, name, transitions):
+        self._name = name
+        self._transitions = transitions
+        self._all_source_states = [*set([item for sublist in transitions.values() for item in sublist])]
+        self._all_target_states = [item for item in transitions.keys()]
+
+    def validate_transition(self, prev_state):
+        if prev_state in self._all_target_states:
+            logger.info(f'Resource is already in target state ({prev_state}) in {self._all_target_states}')
+            return False
+        elif prev_state not in self._all_source_states:
+            raise exceptions.UnauthorizedOperation(
+                action=self._name,
+                message=f'This transition is not possible, {prev_state} cannot go to {self._all_target_states}. '
+                f'If there is a sharing or revoking in progress wait until it is complete and try again.',
+            )
+        else:
+            return True
+
+    def get_transition_target(self, prev_state):
+        if self.validate_transition(prev_state):
+            for target_state, list_prev_states in self._transitions.items():
+                if prev_state in list_prev_states:
```

Comment:
nit: `not in` and drop the else branch

Reply:
I just copy/pasted the code. I was very tempted to do changes on the implementation of the state machines, but I think we should do a review of them in a separate PR